src/adaptivity.jl

```diff
@@ -0,0 +1,388 @@
+            info = ReturnCode.Failure
+            mesh_new = Nothing
+        else
+            #println("Mesh redistributing")
+            mesh_new = redistribute(mesh_current, n, Nsub_star, s_hat)
+        end
+    end
+    return mesh_new, Nsub_star, info
+end
+
+"""
+    redistribute(mesh_current, n, Nsub_star, s_hat)
+
+Generate a new mesh.
+"""
+function redistribute(mesh_current::Vector,
+    n::Int64,
+    Nsub_star::Int64,
+    s_hat::Vector{Float64})
+    mesh_new = zeros(Float64, Nsub_star + 1)
+    sum = 0.0
+    for k in 1:n
+        sum += s_hat[k] * (mesh_current[k + 1] - mesh_current[k])
+    end
+    zeta = sum / Nsub_star
+    k::Int64 = 1
+    i::Int64 = 0
+    mesh_new[1] = mesh_current[1]
+    t = mesh_current[1]
+    integral::Float64 = 0.0
+    while k <= n
+        next_piece = s_hat[k] * (mesh_current[k + 1] - t)
+        if (integral + next_piece) > zeta
+            mesh_new[i + 2] = (zeta - integral) / s_hat[k] + t
+            t = mesh_new[i + 2]
+            i += 1
+            integral = 0
+        else
+            integral += next_piece
+            t = mesh_current[k + 1]
+            k += 1
+        end
+    end
+    mesh_new[Nsub_star + 1] = mesh_current[n + 1]
+    return mesh_new
+end
+
+"""
+    half_mesh(mesh_current)
+
+The input mesh_current has length of n+1
+
+Divide the original subinterval into two equal length subinterval.
+"""
+function half_mesh(mesh_current::Vector)
+    n = length(mesh_current) - 1
+    mesh_new = zeros(Float64, 2 * n + 1)
+    mesh_new[1] = mesh_current[1]
+    for i in 1:n
+        mesh_new[2 * i + 1] = mesh_current[i + 1]
+        mesh_new[2 * i] = (mesh_current[i + 1] + mesh_current[i]) / 2.0
+    end
+    return mesh_new
+end
+
+function idamax(x)
+    x = abs.(x)
+    _, id = findmax(x)
+    return id
+end
+
+"""
+    defect_estimate(prob, Y, alg, n, dt, mesh, k_discrete)
+
+defect_estimate use the discrete solution approximation Y, plus stages of 
+the RK method in 'k_discrete', plus some new stages in 'k_interp' to construct 
+an interpolant
+"""
+function defect_estimate(S::BVPSystem,
+    cache::AbstractMIRKCache,
+    alg::Union{GeneralMIRK, MIRK},
+    TU::MIRKTableau)
+    n, len, Y, p, k_discrete, mesh, f = S.N - 1,
+    S.M,
+    S.y,
+    S.p,
+    cache.k_discrete,
+    S.x,
+    S.fun!
+    s, s_star, tau_star = TU.s, TU.s_star, TU.tau
+
+    # Initialization
+    defect = zeros(n, len)
+    #s, s_star, tau_star, x_star, v_star, c_star = setup_coeff(alg)
+
+    f_sample_1, f_sample_2 = zeros(Float64, len), zeros(Float64, len)
+    def_1, def_2 = zeros(Float64, len), zeros(Float64, len)
+    temp_1, temp_2 = zeros(Float64, len), zeros(Float64, len)
+    estimate_1, estimate_2 = zeros(Float64), zeros(Float64)
+
+    # Evaluate at the first sample point
+    weights_1, weights_1_prime = interp_weights(tau_star, alg)
+    # Evaluate at the second sample point
+    weights_2, weights_2_prime = interp_weights(1.0 - tau_star, alg)
```

Comment:
```suggestion
    weights_2, weights_2_prime = interp_weights(1 - tau_star, alg)
```

etc. shouldn't assume float64

Reply:
I think while we can know the type of ```tau_star```, why don't we directly use Float to avoid abundant type conversion?